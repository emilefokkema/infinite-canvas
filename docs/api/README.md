ef-infinite-canvas

# ef-infinite-canvas

## Enumerations

- [Units](api/enums/Units.md)

## Interfaces

- [Config](api/interfaces/Config.md)
- [DrawEvent](api/interfaces/DrawEvent.md)
- [EventMap](api/interfaces/EventMap.md)
- [InfiniteCanvas](api/interfaces/InfiniteCanvas.md)
- [InfiniteCanvasCtr](api/interfaces/InfiniteCanvasCtr.md)
- [InfiniteCanvasEventHandlers](api/interfaces/InfiniteCanvasEventHandlers.md)
- [InfiniteCanvasEventWithDefaultBehavior](api/interfaces/InfiniteCanvasEventWithDefaultBehavior.md)
- [InfiniteCanvasRenderingContext2D](api/interfaces/InfiniteCanvasRenderingContext2D.md)
- [InfiniteCanvasTouch](api/interfaces/InfiniteCanvasTouch.md)
- [InfiniteCanvasTouchEvent](api/interfaces/InfiniteCanvasTouchEvent.md)
- [InfiniteCanvasTouchList](api/interfaces/InfiniteCanvasTouchList.md)
- [TransformationEvent](api/interfaces/TransformationEvent.md)
- [TransformationRepresentation](api/interfaces/TransformationRepresentation.md)
- [Transformed](api/interfaces/Transformed.md)

## Variables

### default

â€¢ **default**: [`InfiniteCanvasCtr`](api/interfaces/InfiniteCanvasCtr.md)

#### Defined in

[infinite-canvas.ts:49](https://github.com/emilefokkema/infinite-canvas/blob/4a1afe1/src/api-surface/infinite-canvas.ts#L49)
