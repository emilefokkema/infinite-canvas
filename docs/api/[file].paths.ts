import { loadPaths } from '../.vitepress/api-docs/load-paths'

export default {
    async paths(){
        if(process.env.NODE_ENV !== 'production'){
            return [];
        }
        return await loadPaths();
    }
}