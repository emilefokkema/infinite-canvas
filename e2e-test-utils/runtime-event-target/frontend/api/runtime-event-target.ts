import { ChartMap } from '../../shared/serializable-types'

export interface RuntimeEventTarget<TMap>{
    emitEvents(map: ChartMap<TMap>): void
    handleEvents<TType extends keyof TMap>(type: TType, handler: (e: TMap[TType]) => void): void
    switchToCapture(type: keyof TMap): void
}