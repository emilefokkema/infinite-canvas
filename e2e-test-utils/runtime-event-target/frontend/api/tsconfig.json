{
    "compilerOptions": {
        "module": "ESNext",
        "moduleResolution": "bundler",
        "target": "ESNext",
        "composite": true,
        "emitDeclarationOnly": true,
        "outDir": "./dist",
        "lib": ["DOM"]
    },
    "references": [
        {"path": "../../shared"}
    ],
    "include": ["**/*"]
}