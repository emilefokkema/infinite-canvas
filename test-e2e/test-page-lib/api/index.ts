import type { InfiniteCanvas } from '../../../src/api-surface/infinite-canvas'
import { CanvasElementInitialization, InfiniteCanvasInitialization } from './configuration';
import { CanvasElementWrapper, AttachedEventListener } from './wrappers';
import { EventTarget } from './event-target'

declare global{
    interface Window{
        TestPageLib: TestPageLib
        TestCaseLib: TestCaseLib
    }
}

export const EVENT_LISTENER_DATA = 'EVENT_LISTENER_DATA'

export interface TestPageLib{
    initializeCanvasElement(config: CanvasElementInitialization): CanvasElementWrapper
    initializeInfiniteCanvas(config: CanvasElementInitialization & InfiniteCanvasInitialization): Promise<InfiniteCanvas>
    openMessagePort(url: string): Promise<void>;
    addEventListener<
        TEventName,
        TEventType,
        TEventTarget extends EventTarget<TEventType, TEventName>
        >(target: TEventTarget, name: TEventName, capture: boolean | undefined, eventListenerId: string): AttachedEventListener<TEventType>
    makeSerializableTextMetrics(textMetrics: TextMetrics): TextMetrics
}

export interface TestCaseLib{
    drawTestCase(testCaseFile: string): Promise<void>
}

export type { CanvasElementInitialization, InfiniteCanvasInitialization, CanvasElementWrapper, EventTarget, AttachedEventListener }