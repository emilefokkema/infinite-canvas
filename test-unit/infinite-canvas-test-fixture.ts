import { Mock, MockInstance, vi } from 'vitest'
import { DrawingLock } from "src/drawing-lock";
import { MockCanvasMeasurementProvider } from './mock-canvas-measurement-provider';
import { CanvasContextMock } from './canvas-context-mock';
import { Config } from 'api/config';
import { MockDrawingIterationProvider } from './mock-drawing-iteration-provider';
import { CssLengthConverterFactory } from 'src/css-length-converter-factory';
import { InfiniteCanvasViewBox } from 'src/infinite-canvas-viewbox';
import { RectangleManagerImpl } from 'src/rectangle/rectangle-manager-impl';
import { InfiniteContext } from 'src/infinite-context/infinite-context';
import { ViewBox } from 'src/interfaces/viewbox';
import { InfiniteCanvasRenderingContext2D } from 'api/infinite-canvas-rendering-context-2d';


function setupGlobals(){
	window.createImageBitmap = function(): Promise<ImageBitmap>{return undefined;};
	(<any>window).ImageData = class {
		public width: number;
		public height: number;
		public data: Uint8ClampedArray;
		constructor(arrayOrWidth: Uint8ClampedArray | number, widthOrHeight: number, height?: number){
			if(typeof arrayOrWidth === "number"){
				this.width = arrayOrWidth;
				this.height = widthOrHeight;
			}else{
				this.width = widthOrHeight;
				this.height = height;
				this.data = arrayOrWidth;
			}
		}
	};
}

export interface InfiniteCanvasTestFixture{
    width: number
    height: number
    releaseDrawingLockSpy: MockInstance<()=>void>
    getDrawingLockSpy: MockInstance<()=>DrawingLock>
    measurementProvider: MockCanvasMeasurementProvider
    contextMock: CanvasContextMock
    config: Partial<Config>
    mockDrawingIterationProvider: MockDrawingIterationProvider
    viewbox: ViewBox,
    infiniteContext: InfiniteCanvasRenderingContext2D
}

export function createInfiniteCanvasTestFixture(): InfiniteCanvasTestFixture{
    setupGlobals();
    const drawingLock: DrawingLock = {release(){}};
    const releaseDrawingLockSpy = vi.spyOn(drawingLock, 'release');
    const getDrawingLockSpy = vi.fn<()=>DrawingLock>().mockReturnValue(drawingLock)
    const width = 200;
    const height = 200;
    const measurementProvider = new MockCanvasMeasurementProvider(width, height);
    const contextMock = new CanvasContextMock();
    const config = {}
    const mockDrawingIterationProvider = new MockDrawingIterationProvider()
    const cssLengthConverterFactory: CssLengthConverterFactory = {
        create: () => ({
            getNumberOfPixels(value: number){
                return value;
            }
        })
    };
    const viewbox = new InfiniteCanvasViewBox(
        new RectangleManagerImpl(measurementProvider, config),
        contextMock.mock,
        mockDrawingIterationProvider,
        getDrawingLockSpy,
        () => false);
    const infiniteContext = new InfiniteContext(undefined, viewbox, cssLengthConverterFactory);
    return {
        width,
        height,
        releaseDrawingLockSpy,
        getDrawingLockSpy,
        measurementProvider,
        contextMock,
        config,
        mockDrawingIterationProvider,
        viewbox,
        infiniteContext
    };
}