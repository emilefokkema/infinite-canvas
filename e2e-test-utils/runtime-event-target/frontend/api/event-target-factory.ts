import { EventTargetLike } from '../../shared/event-target-like'
import { ConnectionData } from '../../shared/connection-data';
import { RuntimeEventTarget } from './runtime-event-target'

export interface EventTargetFactory{
    createEventTarget<TMap>(
        eventTargetLike: EventTargetLike<TMap>,
        connection: ConnectionData
    ): RuntimeEventTarget<TMap>
}