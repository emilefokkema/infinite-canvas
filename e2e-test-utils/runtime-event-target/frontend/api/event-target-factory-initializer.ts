import { EventTargetFactory } from "./event-target-factory";

export type EventTargetFactoryInitializer = () => EventTargetFactory