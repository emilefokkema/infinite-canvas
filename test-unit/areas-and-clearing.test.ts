import { beforeEach, describe, expect, it} from 'vitest'
import {CanvasContextMock} from "./canvas-context-mock";
import {InfiniteCanvasRenderingContext2D} from "api/infinite-canvas-rendering-context-2d";
import { createInfiniteCanvasTestFixture } from './infinite-canvas-test-fixture';

describe('an infinite canvas context', () => {
	let infiniteContext: InfiniteCanvasRenderingContext2D;
	let contextMock: CanvasContextMock;

    beforeEach(() => {
		({contextMock, infiniteContext} = createInfiniteCanvasTestFixture());
	});

    describe("that makes a path with zero area and strokes it", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.moveTo(20, 20);
			infiniteContext.lineTo(30, 30);
			infiniteContext.stroke();
		});

		describe("and then clears an area overlapping but not covering the path", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(25, 25, 30, 30);
			});

			it("should have added a clearRect", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe("that makes a path, fills a rect, fills the path, begins a new path and then clears a rect containing the first drawn path", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.moveTo(0, 0);
			infiniteContext.lineTo(1, 0);
			infiniteContext.lineTo(1, 1);
			infiniteContext.lineTo(0, 1);
			infiniteContext.lineTo(0, 0);
			infiniteContext.fillRect(3, 3, 1, 1);
			infiniteContext.fill();
			infiniteContext.beginPath();
			contextMock.clear();
			infiniteContext.clearRect(-1, -1, 3, 3);
		});

		it("should forget the first drawn path", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that draws a triangular path and then clears a rect outside the triangle", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
			infiniteContext.moveTo(0, 0);
			infiniteContext.lineTo(4, 0);
			infiniteContext.lineTo(0, 4);
			infiniteContext.fill();
			contextMock.clear();
			infiniteContext.clearRect(3, 3, 2, 2);
		});

		it("should not have added a clearRect", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rect and adds a clearRect that partially covers it", () => {

		beforeEach(() => {
			infiniteContext.fillRect(50, 50, 50, 50);
			infiniteContext.clearRect(40, 40, 20, 20);
		});

		describe("and then adds a clearRect with negative width and height that covers the previous clearRect entirely", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(75, 75, -50, -50);
			});

			it("should end up with only one clearRect", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe("that begins path", () => {

		beforeEach(() => {
			infiniteContext.beginPath();
		});

		describe("and then changes state and fills a rect", () => {

			beforeEach(() => {
				infiniteContext.fillStyle = "#f00";
				contextMock.clear();
				infiniteContext.fillRect(0, 0, 2, 2);
			});

			it("should have remembered the state change", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});

			describe("and then adds a rect to the path and fills it", () => {

				beforeEach(() => {
					infiniteContext.rect(0, 2, 2, 2);
					contextMock.clear();
					infiniteContext.fill();
				});

				it("should do that using the same changed state", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});
		});

		describe("and then changes state, begins a new path and fills it", () => {

			beforeEach(() => {
				infiniteContext.fillStyle = "#f00";
				infiniteContext.beginPath();
				infiniteContext.rect(0, 0, 2, 2);
				contextMock.clear();
				infiniteContext.fill();
			});

			it("should have remembered the state change", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});

		describe("and then builds it and fills it", () => {

			beforeEach(() => {
				infiniteContext.moveTo(0,0);
				infiniteContext.lineTo(3, 0);
				infiniteContext.lineTo(0, 3);
				infiniteContext.closePath();
				contextMock.clear();
				infiniteContext.fill();
			});

			it("should have executed the new instructions", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});

			describe("and then clears an area that is outside the drawn area", () => {

				beforeEach(() => {
					contextMock.clear();
					infiniteContext.clearRect(10, 10, 1, 1);
				});

				it("should not have done anything", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});
			});

			describe("and then clears a smaller area than the one that was closed and adds another instruction", () => {

				beforeEach(() => {
					infiniteContext.clearRect(0, 0, 2, 2);
					infiniteContext.strokeStyle = "#f00";
					infiniteContext.beginPath();
					infiniteContext.moveTo(0,0);
					infiniteContext.lineTo(2, 0);
					contextMock.clear();
					infiniteContext.stroke();
				});

				it("should still have executed the instructions in the completed area and should have added a clear rect instruction", () => {
					expect(contextMock.getLog()).toMatchSnapshot();
				});

				describe("and then clears an area containing all previous instructions", () => {

					beforeEach(() => {
						contextMock.clear();
						infiniteContext.clearRect(-1, -1, 4, 4);
					});

					it("should have cleared a rectangle once", () => {
						expect(contextMock.getLog()).toMatchSnapshot();
					});

					describe("and then draws something else", () => {

						beforeEach(() => {
							contextMock.clear();
							infiniteContext.fillRect(0, 0, 1, 1);
						});

						it("should have cleared a rectangle once more", () => {
							expect(contextMock.getLog()).toMatchSnapshot();
						});
					});
				});
			});
			
		});
	});

    describe("that alters its state and draws a rectangle", () => {

		beforeEach(() => {
			infiniteContext.strokeStyle = "#f00";
			infiniteContext.fillStyle = "#00f";
			contextMock.clear();
			infiniteContext.fillRect(1,1,1,1);
		});

		it("should have called the context methods", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});

		describe("and then clears a rectangle containing the drawing", () => {

			beforeEach(() => {
				contextMock.clear();
				infiniteContext.clearRect(0, 0, 3, 3);
			});

			it("should have cleared a rectangle and nothing else", () => {
				expect(contextMock.getLog()).toMatchSnapshot();
			});
		});
	});

    describe("that fills a rectangle, fills a smaller rectangle inside it and clears a rectangle containing the small one but not the big one", () => {

		beforeEach(() => {
			infiniteContext.fillRect(0, 0, 5, 5);
			infiniteContext.fillRect(2, 2, 1, 1);
			contextMock.clear();
			infiniteContext.clearRect(1, 1, 3, 3);
		});

		it("should forget all about the second rectangle", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

    describe("that fills a rectangle, clears part of it and then clears a bigger part of it", () => {

		beforeEach(() => {
			infiniteContext.fillRect(0, 0, 5, 5);
			infiniteContext.clearRect(2, 2, 1, 1);
			contextMock.clear();
			infiniteContext.clearRect(1, 1, 3, 3);
		});

		it("should end up with only one clear rect instruction", () => {
			expect(contextMock.getLog()).toMatchSnapshot();
		});
	});

})