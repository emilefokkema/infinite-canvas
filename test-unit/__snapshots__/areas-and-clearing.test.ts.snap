// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`an infinite canvas context > that alters its state and draws a rectangle > and then clears a rectangle containing the drawing > should have cleared a rectangle and nothing else 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
]
`;

exports[`an infinite canvas context > that alters its state and draws a rectangle > should have called the context methods 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#00f"",
  "context.lineWidth = 1",
  "context.strokeStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(1,1)",
  "context.lineTo(2,1)",
  "context.lineTo(2,2)",
  "context.lineTo(1,2)",
  "context.closePath()",
  "context.moveTo(1,1)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that begins path > and then builds it and fills it > and then clears a smaller area than the one that was closed and adds another instruction > and then clears an area containing all previous instructions > and then draws something else > should have cleared a rectangle once more 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.strokeStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(1,0)",
  "context.lineTo(1,1)",
  "context.lineTo(0,1)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that begins path > and then builds it and fills it > and then clears a smaller area than the one that was closed and adds another instruction > and then clears an area containing all previous instructions > should have cleared a rectangle once 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
]
`;

exports[`an infinite canvas context > that begins path > and then builds it and fills it > and then clears a smaller area than the one that was closed and adds another instruction > should still have executed the instructions in the completed area and should have added a clear rect instruction 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(3,0)",
  "context.lineTo(0,3)",
  "context.closePath()",
  "context.lineWidth = 1",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(0,0,2,2)",
  "context.restore()",
  "context.strokeStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.stroke()",
]
`;

exports[`an infinite canvas context > that begins path > and then builds it and fills it > and then clears an area that is outside the drawn area > should not have done anything 1`] = `[]`;

exports[`an infinite canvas context > that begins path > and then builds it and fills it > should have executed the new instructions 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(3,0)",
  "context.lineTo(0,3)",
  "context.closePath()",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that begins path > and then changes state and fills a rect > and then adds a rect to the path and fills it > should do that using the same changed state 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.fillStyle = "#000"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.fillStyle = "#f00"",
  "context.moveTo(0,2)",
  "context.lineTo(2,2)",
  "context.lineTo(2,4)",
  "context.lineTo(0,4)",
  "context.closePath()",
  "context.moveTo(0,2)",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that begins path > and then changes state and fills a rect > should have remembered the state change 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that begins path > and then changes state, begins a new path and fills it > should have remembered the state change 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.fillStyle = "#f00"",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(2,0)",
  "context.lineTo(2,2)",
  "context.lineTo(0,2)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.lineWidth = 1",
  "context.fill()",
]
`;

exports[`an infinite canvas context > that draws a triangular path and then clears a rect outside the triangle > should not have added a clearRect 1`] = `[]`;

exports[`an infinite canvas context > that fills a rect and adds a clearRect that partially covers it > and then adds a clearRect with negative width and height that covers the previous clearRect entirely > should end up with only one clearRect 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(50,50)",
  "context.lineTo(100,50)",
  "context.lineTo(100,100)",
  "context.lineTo(50,100)",
  "context.closePath()",
  "context.moveTo(50,50)",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(75,75,-50,-50)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that fills a rectangle, clears part of it and then clears a bigger part of it > should end up with only one clear rect instruction 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(5,0)",
  "context.lineTo(5,5)",
  "context.lineTo(0,5)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(1,1,3,3)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that fills a rectangle, fills a smaller rectangle inside it and clears a rectangle containing the small one but not the big one > should forget all about the second rectangle 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(0,0)",
  "context.lineTo(5,0)",
  "context.lineTo(5,5)",
  "context.lineTo(0,5)",
  "context.closePath()",
  "context.moveTo(0,0)",
  "context.fill()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(1,1,3,3)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that makes a path with zero area and strokes it > and then clears an area overlapping but not covering the path > should have added a clearRect 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.beginPath()",
  "context.moveTo(20,20)",
  "context.lineTo(30,30)",
  "context.lineWidth = 1",
  "context.stroke()",
  "context.save()",
  "context.transform(1,0,0,1,0,0)",
  "context.clearRect(25,25,30,30)",
  "context.restore()",
]
`;

exports[`an infinite canvas context > that makes a path, fills a rect, fills the path, begins a new path and then clears a rect containing the first drawn path > should forget the first drawn path 1`] = `
[
  "context.restore()",
  "context.save()",
  "context.clearRect(0,0,200,200)",
  "context.lineWidth = 1",
  "context.beginPath()",
  "context.moveTo(3,3)",
  "context.lineTo(4,3)",
  "context.lineTo(4,4)",
  "context.lineTo(3,4)",
  "context.closePath()",
  "context.moveTo(3,3)",
  "context.fill()",
]
`;
